content/browser/webui/brave_shared_resources_data_source.cc

```diff
@@ -0,0 +1,216 @@
+}
+
+void AddResourcesToMap(ResourcesMap* resources_map) {
+  const std::map<std::string, std::string> aliases =
+      CreatePathPrefixAliasesMap();
+
+  for (size_t i = 0; i < kBraveWebuiResourcesSize; ++i) {
+    const auto& resource = kBraveWebuiResources[i];
+
+    AddResource(resource.name, resource.value, resource.gzipped, resources_map);
+
+    for (auto it = aliases.begin(); it != aliases.end(); ++it) {
+      if (base::StartsWith(resource.name, it->first,
+                           base::CompareCase::SENSITIVE)) {
+        std::string resource_name(resource.name);
+        AddResource(it->second + resource_name.substr(it->first.length()),
+                    resource.value, resource.gzipped, resources_map);
+      }
+    }
+  }
+}
+
+const ResourcesMap* CreateResourcesMap() {
+  ResourcesMap* result = new ResourcesMap();
+  AddResourcesToMap(result);
+  return result;
+}
+
+const ResourcesMap& GetResourcesMap() {
+  // This pointer will be intentionally leaked on shutdown.
+  static const ResourcesMap* resources_map = CreateResourcesMap();
+  return *resources_map;
+}
+
+int GetIdrForPath(const std::string& path) {
+  const ResourcesMap& resources_map = GetResourcesMap();
+  auto it = resources_map.find(path);
+  return it != resources_map.end() ? it->second.idr : -1;
+}
+
+}  // namespace
+
+BraveSharedResourcesDataSource::BraveSharedResourcesDataSource() {
+}
+
+BraveSharedResourcesDataSource::~BraveSharedResourcesDataSource() {
+}
+
+std::string BraveSharedResourcesDataSource::GetSource() const {
+  return "brave-resources";
+}
+
+void BraveSharedResourcesDataSource::StartDataRequest(
+    const std::string& path,
+    const ResourceRequestInfo::WebContentsGetter& wc_getter,
+    const URLDataSource::GotDataCallback& callback) {
+  int idr = GetIdrForPath(path);
+  DCHECK_NE(-1, idr) << " path: " << path;
+  scoped_refptr<base::RefCountedMemory> bytes;
+
+  // Cannot access GetContentClient() from here as that is //content/public only.
+  // Therefore cannot access ContentClient::GetDataResourceBytes,
+  // so go to the bundle directly.
+  // This will work for all content clients apart from in a test environment,
+  // where this shoudl be mocked.
+  bytes = ui::ResourceBundle::GetSharedInstance().LoadDataResourceBytes(idr);
+  callback.Run(bytes.get());
+}
+
+bool BraveSharedResourcesDataSource::AllowCaching() const {
+  // Should not be cached to reflect dynamically-generated contents that may
+  // depend on the current locale.
+  return true;
+}
+
+std::string BraveSharedResourcesDataSource::GetMimeType(
+    const std::string& path) const {
+  if (path.empty())
+    return "text/html";
+
+#if defined(OS_WIN)
+  base::FilePath file(base::UTF8ToWide(path));
+  std::string extension = base::WideToUTF8(file.FinalExtension());
+#else
+  base::FilePath file(path);
+  std::string extension = file.FinalExtension();
+#endif
+
+  if (!extension.empty())
+    extension.erase(0, 1);
+
+  if (extension == "html")
```

Comment:
I know these resources being served are ones we own- @diracdeltas any concern about deriving the mime type strictly from the extension?

Reply:
thanks for bringing this up. if these are local files that an attacker can't modify or rename, i think this is ok.

Reply:
the alternative would be to hard-code a mapping of file names to MIME type and refuse to load any files that don't have one of the whitelisted file names

Reply:
this is modeled on the existing chromium `SharedResourcesDataSource` which has the same mime type logic. It works for us too because it only serves files we bundle, and those files are correctly suffixed.